cdc/log.cc

```diff
@@ -539,12 +539,109 @@ class transformer final {
+            });
+            if (i != res.end() && i->first == k) {
+                i->second = v;
+            } else {
+                res.emplace(i, k, v);
+            }
+        });
+        set_for_each(deleted, [&res](bytes_view k) {
+            auto i = std::find_if(res.begin(), res.end(), [&k](std::pair<bytes_view, bytes_view>& p) {
+                return p.first == k;
+            });
+            if (i != res.end()) {
+                res.erase(i);
+            }
+        });
+        return map_type_impl::serialize_partially_deserialized_form(res, cql_serialization_format::internal());
+    }
+    static bytes merge_sets(const abstract_type& type, const bytes_opt& prev, const bytes_opt& next, const bytes_opt& deleted) {
+        std::vector<bytes_view> res;
+        set_for_each(prev, [&res](bytes_view k) {
+            res.emplace_back(k);
+        });
+        set_for_each(next, [&res, &type](bytes_view k) {
+            auto i = std::find_if(res.begin(), res.end(), [&k, &type](bytes_view rk) {
+                return type.compare(rk, k) >= 0;
+            });
+            if (i == res.end() || *i != k) {
+                res.emplace(i, k);
+            }
+        });
+        set_for_each(deleted, [&res](bytes_view k) {
+            auto i = std::find_if(res.begin(), res.end(), [&k](bytes_view rk) {
+                return rk == k;
+            });
+            if (i != res.end()) {
+                res.erase(i);
+            }
+        });
+        return set_type_impl::serialize_partially_deserialized_form(res, cql_serialization_format::internal());
+    }
+    static bytes merge_udt(const user_type_impl& type, const bytes_opt& prev, const bytes_opt& next, const bytes_opt& deleted) {
+        std::vector<bytes_view_opt> res(type.size());
+        udt_for_each(prev, [&res, i = res.begin()](bytes_view_opt k) mutable {
+            *i++ = k;
+        });
+        udt_for_each(next, [&res, i = res.begin()](bytes_view_opt k) mutable {
+            if (k) {
+                *i = k;
+            }
+            ++i;
+        });
+        set_for_each(deleted, [&res](bytes_view k) {
+            auto index = deserialize_field_index(k);
+            res[index] = std::nullopt;
+        });
+        return type.build_value(res);
+    }
+
     // TODO: is pre-image data based on query enough. We only have actual column data. Do we need
     // more details like tombstones/ttl? Probably not but keep in mind.
     mutation transform(const mutation& m, const cql3::untyped_result_set* rs, api::timestamp_type ts, bytes tuuid, int& batch_no) const {
         auto stream_id = _ctx._cdc_metadata.get_stream(ts, m.token());
         mutation res(_log_schema, stream_id.to_partition_key(*_log_schema));
-
+        // should we generate post-image data? note: we can only do this if both pre- and postimage 
+        // is enabled.
+        const auto postimage = _schema->cdc_options().preimage() && _schema->cdc_options().postimage();
```

Comment:
This is inconsistent with line 988/1140. There you fetch preimage when postimage is set to true no matter whether preimage is set to true or not. Why is there OR in that line and AND in this?

Reply:
Oops. Mistake. First draft thing.